### DISCUSSION TITLE
Other Fortran static analysis tools

### DISCUSSION THREAD

**Discussion by jmunroe** [marshallward/flint#9] [Ideas]
For possible inclusion into some future documentation:

From section 6.2 Luca Bonaventura, René Redler and Reinhard Budich. “Earth System Modelling - Volume 2.”, there is a section on "Software Coding Standards".  It mentions the following Fortran static analyzers:

“FLINT Manual (1994) User’s manual, FORTRAN-lint source code analyzer. http://www.fnal.gov/docs/products/flint/manual.txt”
- does this still exist?

“FTNchek Tool (2004) ftnchek static analyzer for Fortran. http://www.dsm.fordham.edu/ftnchek/”
- f77 only?

A quick Google search leads to a commercial product:

Cleanscape FortranLint
https://stellar.cleanscape.net/products/fortranlint/index.html

**Comment by marshallward** [comment]
Another possibility here is the LLVM-based compilers, which seem to provide more details about the AST.

https://github.com/flang-compiler/flang

I am not certain, but I think this is what LFortran (also Python-based) uses in its Fortran REPL:

https://lfortran.org/

That said, I think there's still value in having well-defined objects which can be inspected and manipulated in Python.  Also, I strongly suspect that the LLVM compilers are chucking out all the whitespace and comments and other things which we probably care more about.

**Comment by marshallward** [comment]
Courtesy of Peter Teuben (@teuben): another "flint" capable of compile-time analysis.

https://github.com/astroumd/miriad/blob/master/src/tools/flint.c

https://github.com/marshallward/flint/issues/12

**Comment by marshallward** [comment]
Cerfacs has a Fortran linter, even written in Python.

  https://pypi.org/project/flinter/

I couldn't get its website to load, so couldn't learn much more.

It appears to be regex-based (an approach I had to abandon at some stage, since I found it difficult to generate tokens).

It seems to enforce these guidelines: https://alm.engr.colostate.edu/cb/wiki/16983

**Comment by marshallward** [comment]
[fpt from SimCon](http://www.simconglobal.com/fpt_summary.html)

Saw a nice presentation on this product.  Although not open-source, it appears to been developed through collaboration with universities and government in the UK (if I understood it correctly) and may be available for wider use under the right circumstances.
